src/coreclr/jit/lclmorph.cpp

```diff
@@ -678,16 +678,52 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
 
         // In general we don't know how an exposed struct field address will be used - it may be used to
         // access only that specific field or it may be used to access other fields in the same struct
-        // be using pointer/ref arithmetic. It seems reasonable to make an exception for the "this" arg
-        // of calls - it would be highly unsual for a struct member method to attempt to access memory
+        // by using pointer/ref arithmetic. It seems reasonable to make an exception for the "this" arg
+        // of calls - it would be highly unusual for a struct member method to attempt to access memory
         // beyond "this" instance. And calling struct member methods is common enough that attempting to
         // mark the entire struct as address exposed results in CQ regressions.
-        bool isThisArg = user->IsCall() && (user->AsCall()->gtCallThisArg != nullptr) &&
-                         (val.Node() == user->AsCall()->gtCallThisArg->GetNode());
+        GenTreeCall* callTree  = user->IsCall() ? user->AsCall() : nullptr;
+        bool         isThisArg = (callTree != nullptr) && (callTree->gtCallThisArg != nullptr) &&
+                         (val.Node() == callTree->gtCallThisArg->GetNode());
         bool exposeParentLcl = varDsc->lvIsStructField && !isThisArg;
 
-        m_compiler->lvaSetVarAddrExposed(exposeParentLcl ? varDsc->lvParentLcl
-                                                         : val.LclNum() DEBUGARG(AddressExposedReason::ESCAPE_ADDRESS));
+        bool hasHiddenStructArg = false;
+        if (m_compiler->opts.compJitOptimizeStructHiddenBuffer)
+        {
+            if (varTypeIsStruct(varDsc) && varDsc->lvIsTemp)
+            {
+                if ((callTree != nullptr) && callTree->HasRetBufArg() && val.Node() == callTree->gtCallArgs->GetNode())
+                {
+                    // Mark the local variable associated as being defined at this position.
+                    assert(!exposeParentLcl);
+                    assert(val.Node()->TypeIs(TYP_BYREF, TYP_I_IMPL));
+
+                    GenTree* addr = val.Node();
+                    assert(addr->OperIs(GT_ADDR));
+                    GenTree* hiddenArgOp = addr->AsOp()->gtGetOp1();
+
+                    if (hiddenArgOp->OperIs(GT_FIELD))
```

Comment:
My expectation would be that we can have an arbitrary long sequence of fields here, e. g.:
```cs
private static long Problem()
{
    var a = default(LargeStruct);

    a.SmallerStruct1.MultiRegStruct1 = GetMultiRegStruct();

    return a.SmallerStruct2.MultiRegStruct1.FirstLngValue;
}
```
```scala
[000008] S-C-G-------              *  CALL      void   RyuJitReproduction.Program.GetMultiRegStruct
[000010] ----G------- arg0         \--*  ADDR      byref 
[000009] ----G--N----                 \--*  FIELD     struct MultiRegStruct1
[000007] ------------                    \--*  ADDR      byref 
[000006] -------N----                       \--*  FIELD     struct SmallerStruct1
[000005] ------------                          \--*  ADDR      byref 
[000004] -------N----                             \--*  LCL_VAR   struct<LargeStruct, 96> V00 loc0         
```